### PR TITLE
Code moved from amp/http-client PR #270

### PR DESCRIPTION
**Pull request by remorhaz**
Just moved code from [PR#270](https://github.com/amphp/http-client/pull/270). The original PR may be rejected, I guess.

### CODE REVIEW

File: examples/basic.php [hunk 1]
```diff
@@ -1,21 +1,46 @@
...
 
-    /** @var Response $secondResponse */
-    $secondResponse = yield $httpClient->request(new Request('https://google.com/'));
-    yield $secondResponse->getBody()->buffer();
-});
+        $firstPsrRequest = $psrRequestFactory->createRequest('GET', 'https://google.com/');
+        /** @var Request $firstAmpRequest */
+        $firstAmpRequest = yield $psrAdapter->fromPsrRequest($firstPsrRequest);
+        // TODO: Investigate if this client bug or if Host header must be cleaned automatically
+        $firstAmpRequest->removeHeader('Host');
```
Comment (remorhaz) [inline examples/basic.php:29]: I have a problem here. Seems that if Amp request contains `Host` header, it redirects in cycle. I don't know if it is a bug or expected behavior of `amp/http-client`, but it looks like a bug.

File: src/PsrAdapter.php [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                return $target;
+            }
+        );
+    }
+
+    /**
+     * @param InputStream     $source
+     * @param StreamInterface $target
+     * @return \Generator
+     * @psalm-return \Generator<mixed,Promise>
```
Comment (enumag) [inline src/PsrAdapter.php:90]:
```suggestion
     * @psalm-return \Generator<mixed, Promise>
```

**Comment by enumag** [comment]
ping @trowski @kelunik

**Comment by kelunik** [comment]
I've manually squashed and merged the PR via https://github.com/amphp/http-client-psr7/commit/487af2548c154c8a1d61bae2692817575b882e39, thanks! I've also changed a few things slightly.

I guess the `PsrAdapter` could be split, but I'm not sure on that.

**Comment by kelunik** [comment]
@remorhaz Thanks for noticing the infinite redirect, it's actually a security issue fixed in https://github.com/amphp/http-client/commit/fa7925363e6d5a0d0d337e2e6eb1affb93cf226e.

**Pull request closed without merging**